fuse_models/src/odometry_3d_publisher.cpp

```diff
@@ -0,0 +1,624 @@
+    acceleration.accel.accel.angular.y = 0.0;
+    acceleration.accel.accel.angular.z = 0.0;
+  } catch (const std::exception & e) {
+    RCLCPP_WARN_STREAM_THROTTLE(
+      logger_, *clock_, 10.0 * 1000,
+      "Failed to find a state at time " << stamp.nanoseconds() << ". Error: " << e.what());
+    return false;
+  } catch (...) {
+    RCLCPP_WARN_STREAM_THROTTLE(
+      logger_, *clock_, 10.0 * 1000,
+      "Failed to find a state at time " << stamp.nanoseconds() << ". Error: unknown");
+    return false;
+  }
+
+  return true;
+}
+
+void Odometry3DPublisher::publishTimerCallback()
+{
+  rclcpp::Time latest_stamp;
+  rclcpp::Time latest_covariance_stamp;
+  bool latest_covariance_valid;
+  nav_msgs::msg::Odometry odom_output;
+  geometry_msgs::msg::AccelWithCovarianceStamped acceleration_output;
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    latest_stamp = latest_stamp_;
+    latest_covariance_stamp = latest_covariance_stamp_;
+    latest_covariance_valid = latest_covariance_valid_;
+    odom_output = odom_output_;
+    acceleration_output = acceleration_output_;
+  }
+
+  if (0u == latest_stamp.nanoseconds()) {
+    RCLCPP_WARN_STREAM_EXPRESSION(
+      logger_, delayed_throttle_filter_.isEnabled(),
+      "No valid state data yet. Delaying tf broadcast.");
+    return;
+  }
+
+  tf2::Transform pose;
+  tf2::fromMsg(odom_output.pose.pose, pose);
+
+  // If requested, we need to project our state forward in time using the 3D kinematic model
+  if (params_.predict_to_current_time) {
+    rclcpp::Time timer_now = interfaces_.get_node_clock_interface()->get_clock()->now();
+
+    // Convert pose in Eigen representation
+    fuse_core::Vector3d position, velocity_linear, velocity_angular;
+    Eigen::Quaterniond orientation;
+    position << pose.getOrigin().x(), pose.getOrigin().y(), pose.getOrigin().z();
+    orientation.x() = pose.getRotation().x();
+    orientation.y() = pose.getRotation().y();
+    orientation.z() = pose.getRotation().z();
+    orientation.w() = pose.getRotation().w();
+    velocity_linear << odom_output.twist.twist.linear.x,
+      odom_output.twist.twist.linear.y, odom_output.twist.twist.linear.z;
+    velocity_angular << odom_output.twist.twist.angular.x,
+      odom_output.twist.twist.angular.y, odom_output.twist.twist.angular.z;
+
+    const double dt = timer_now.seconds() - rclcpp::Time(odom_output.header.stamp).seconds();
+
+    fuse_core::Matrix15d jacobian;
+
+    fuse_core::Vector3d acceleration_linear;
+    if (params_.predict_with_acceleration) {
+      acceleration_linear << acceleration_output.accel.accel.linear.x,
+        acceleration_output.accel.accel.linear.y, acceleration_output.accel.accel.linear.z;
+    }
+
+    predict(
+      position,
+      orientation,
+      velocity_linear,
+      velocity_angular,
+      acceleration_linear,
+      dt,
+      position,
+      orientation,
+      velocity_linear,
+      velocity_angular,
+      acceleration_linear,
+      jacobian);
```

Comment:
`acceleration_linear` should be initialized to zero here, otherwise predict will use an uninitialized value if you aren't predicting with acceleration.